pkg/kubectl/cmd/apiversions.go

```diff
@@ -52,14 +86,62 @@ func RunApiVersions(f *cmdutil.Factory, w io.Writer) error {
 		return err
 	}
 
-	groupList, err := client.Discovery().ServerGroups()
+	o.discovery = client.Discovery()
+	o.out = out
+
+	return nil
+}
+
+// Validate validates all the required options for api-versions.
+func (o *ApiVersionsOptions) Validate() error {
+	return nil
+}
+
+// Run prints out the resources for the specified api version.
+func (o *ApiVersionsOptions) Run() error {
+	groupList, err := o.discovery.ServerGroups()
 	if err != nil {
 		return fmt.Errorf("Couldn't get available api versions from server: %v\n", err)
 	}
 	apiVersions := unversioned_client.ExtractGroupVersions(groupList)
 	sort.Strings(apiVersions)
+
+	w := kubectl.GetNewTabWriter(o.out)
+	defer w.Flush()
+
+	if o.showResources {
+		return o.printResources(apiVersions, w)
+	}
+
+	printGroupVersions(apiVersions, w)
+	return nil
+}
+
+func printGroupVersions(apiVersions []string, w io.Writer) {
 	for _, v := range apiVersions {
 		fmt.Fprintln(w, v)
 	}
+}
+
+func (o *ApiVersionsOptions) printResources(apiVersions []string, w io.Writer) error {
+	if !o.noHeaders {
+		fmt.Fprintln(w, "GROUP/VERSION\tRESOURCE\tNAMESPACED")
+	}
+	for _, groupVersion := range apiVersions {
+		resources, err := o.discovery.ServerResourcesForGroupVersion(groupVersion)
+		if err != nil {
+			return fmt.Errorf("Couldn't get resources from server: %v\n", err)
```

Comment:
> Ok. Should I unwrap in Run() too where it gets ServerGroups?

Yes, throughout.  Because of poor error type identification in go (lack of native chaining is a serious oversight), the barrier to wrapping errors is quite high.